core/include/vecmem/containers/device_vector.ipp

```diff
@@ -0,0 +1,279 @@
+/** VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+// System include(s).
+#include <cassert>
+
+namespace vecmem {
+
+   template< typename TYPE >
+   VECMEM_HOST_AND_DEVICE
+   device_vector< TYPE >::device_vector( device_vector_data< value_type > data )
+   : m_size( data.m_size ), m_ptr( data.m_ptr ) {
+
+   }
+
+   template< typename TYPE >
+   VECMEM_HOST_AND_DEVICE
+   device_vector< TYPE >::device_vector( const device_vector& parent )
+   : m_size( parent.m_size ), m_ptr( parent.m_ptr ) {
+
+   }
+
+   template< typename TYPE >
+   VECMEM_HOST_AND_DEVICE
+   device_vector< TYPE >&
+   device_vector< TYPE >::operator=( const device_vector& rhs ) {
+
+      // Prevent self-assignment.
+      if( this == &rhs ) {
+         return *this;
+      }
+
+      // Copy the other object's payload.
+      m_size = rhs.m_size;
+      m_ptr = rhs.m_ptr;
```

Comment:
I would expect this operation to do a copy, not alias the pointer of the right hand side. Might be confusing to users who are more used to the standard library behaviour. I would suggest documenting this well, or adding a move assignment operator.

Reply:
The behaviour **is** different from STL containers. Since these types do not own the memory that they operate on. As such, a copy can only do this. To create another `vecmem::device_vector` instance that would operate on the same array in memory.

Creating move operators/constructors could be useful, mainly for code documentation purposes. But performance-wise they will not bring us anything...

Reply:
Right, I totally understand the need for this. But semantically I think this class should strictly be constructed and assigned from rvalue vectors. This is actually the exact same discussion we're having a few comments down. :smile:

Reply:
The question is not really the same in the two cases...

The question here is whether we should allow multiple `vecmem::device_vector` objects to operate on the same memory array in the GPU kernel. I believe this can be a useful feature. The N threads of the GPU will already use N `vecmem::device_vector` instances to access the same memory array. What harm does it do if the device code also has multiple instances of the object inside the same thread as well? (Other than allowing for inexperienced developers to produce some "sneakily broken" code.)